pkg/prometheus/watcher.go

```diff
@@ -470,3 +605,105 @@ func (w *Watcher) syncAlertManager(key string) error {
 	_, err = w.mClient.MonitoringV1().Alertmanagers(am.Namespace).Update(am)
 	return err
 }
+
+func (w *Watcher) handleAlertmanagerConfigAdd(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigUpdate(oldObj, newObj interface{}) {
+	key, ok := keyFunc(newObj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config added: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) handleAlertmanagerConfigDelete(obj interface{}) {
+	key, ok := keyFunc(obj)
+	if !ok {
+		return
+	}
+
+	log.Debugf("Alertmanager config deleted: %s", key)
+
+	enqueue(w.amcQueue, key)
+}
+
+func (w *Watcher) syncAlertManagerConfig(key string) error {
+	var alertManagerName string
+	var data []byte
+
+	log.Debugf("syncing alert manager config: key: %s", key)
+	obj, exists, err := w.amcInf.GetIndexer().GetByKey(key)
+	if err != nil {
+		return err
+	}
+
+	if !exists {
+		log.Error("Key not found...")
+		return nil
+	}
+
+	amc := obj.(*monitoringv1.AlertmanagerConfig)
+	if nil == amc {
+		log.Errorf("Got an invalid amc object for: %s", key)
+		return nil
+	}
+
+	log.Debugf("syncing alert manager config: key: %s, name: %s", key, amc.Spec.Type)
+
+	for key, val := range amc.Labels {
+		log.Debugf("Labels: %s %s", key, val)
+		if key == "alertmanager" {
+			alertManagerName = val
+			break
+		}
+	}
+
+	if alertManagerName == "" {
+		log.Errorf("Alert manager label missing in alertmanager config: %s", key)
```

Comment:
this error also seems like a code bug.

Reply:
This is when the AlertmanagerConfig is incorrectly configured and the required label (associating the config object with alert manager instance) is missing. We won't be able to create an appropriate secret without the name, hence the error. Again an assert here will result in pod restart, all we can do is log an error and not create a secret.

Reply:
There is a way to validate kubernetes CRs. Can you file a JIRA to do that? If API validates the CR, it will reject alertmanager configs without valid alertmanager name in it.

Reply:
Created jira: https://platform9.atlassian.net/browse/MA-100

If this the API we can use: https://github.com/kubeflow/crd-validation

Also this will validate the fields in the CR. However we need a label to associate the CR with a valid Alertmanager object. That check (which is the above code) will still be required.